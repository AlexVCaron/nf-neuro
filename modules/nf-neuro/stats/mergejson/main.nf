

process STATS_MERGEJSON {
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_latest.sif':
        'scilus/scilus:latest' }"

    input:
    tuple val(meta), path(jsons)

    output:
    path("*_stats.json")  , emit: json
    path("*_stats.xlsx")  , emit: xlsx
    path "versions.yml"   , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${task.ext.prefix}"
    def suffix = task.ext.suffix ? "${task.ext.suffix}_stats" : "stats"
    def per_point = task.ext.per_point ? true : false
    def average_last_layer = task.ext.average_last_layer ? "--average_last_layer" : ""
    def recursive = task.ext.recursive ? "--recursive" : ""
    def no_list = task.ext.no_list ? "--no_list" : ""
    def stats_over_population = task.ext.stats_over_population ? "--stats_over_population" : ""

    """
    if $per_point;
        then
        for json in $jsons
            do scil_json_merge_entries.py \$json \${json/.json/_avg.json} --remove_parent_key --recursive $average_last_layer
        done
            scil_json_merge_entries.py *_avg.json ${prefix}.json --recursive
    else
        scil_json_merge_entries.py $jsons ${prefix}.json $no_list $recursive
    fi

    scil_json_harmonize_entries.py ${prefix}.json ${prefix}_${suffix}.json -f -v --sort_keys
    scil_json_convert_entries_to_xlsx.py ${prefix}_${suffix}.json ${prefix}_${suffix}.xlsx $stats_over_population

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def suffix = task.ext.first_suffix ? "${task.ext.first_suffix}_stats" : "_stats"
    def prefix = task.ext.prefix ?: "${task.ext.prefix}"

    """
    scil_json_merge_entries.py -h
    scil_json_harmonize_entries.py -h
    scil_json_convert_entries_to_xlsx.py -h

    touch ${prefix}_${suffix}.json
    touch ${prefix}_${suffix}.xlsx

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
