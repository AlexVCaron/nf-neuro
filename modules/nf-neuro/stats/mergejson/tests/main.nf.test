nextflow_process {

    name "Test Process STATS_MERGEJSON"
    script "../main.nf"
    process "STATS_MERGEJSON"

    tag "modules"
    tag "modules_nfcore"
    tag "stats"
    tag "stats/mergejson"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "tractometry.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("stats - length") {
        when {
        config "./nextflow.config"
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/length_stats_*.json")
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.json.get(0).get(1)),
                    file(process.out.xlsx.get(0).get(1)).name,
                    process.out.versions
                ).match() }
            )
        }
    }
    test("stats - merge json -stub-run") {
        tag "stub"
        options "-stub-run"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/length_stats_*.json")
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }
}
