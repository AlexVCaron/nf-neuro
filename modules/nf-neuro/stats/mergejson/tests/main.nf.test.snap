{
    "stats - merge json -stub-run": {
        "content": [
            [
                "versions.yml:md5,4e88428f89f69bbb83a75db27d87a381"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T21:11:08.66582277"
    },
    "stats - length": {
        "content": [
            "length_stats.json:md5,de69902deee6b5da29aa811381c4a2e4",
            "length_stats.xlsx",
            [
                "versions.yml:md5,4e88428f89f69bbb83a75db27d87a381"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T21:10:54.934925518"
    }
}